src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -0,0 +1,493 @@
+import           Foreign.Ptr
+
+{-------------------------------------------------------------------------------
+  Types
+-------------------------------------------------------------------------------}
+
+-- | A k\/op pair with an optional blob reference.
+type KOp = (RawKey, Operation, Maybe BlobRef)
+
+kopKey :: KOp -> RawKey
+kopKey (x,_,_) = x
+
+kopOp :: KOp -> Operation
+kopOp (_,x,_) = x
+
+kopBlobRef :: KOp -> Maybe BlobRef
+kopBlobRef (_,_,x) = x
+
+--
+-- Keys
+--
+
+-- | A string of raw bytes representing a key.
+type RawKey = S.SerialisedKey
+
+-- | Size of key in number of bytes.
+sizeofKey :: RawKey -> Int
+sizeofKey (S.SerialisedKey ba) = sizeofByteArray ba
+
+-- | Size of key in number of bytes.
+sizeofKey16 :: RawKey -> Word16
+sizeofKey16 = fromIntegral . sizeofKey
+
+-- | Size of key in number of bytes.
+sizeofKey64 :: RawKey -> Word64
+sizeofKey64 = fromIntegral . sizeofKey
+
+--
+-- Values
+--
+
+-- | A string of raw bytes representing a value.
+--
+-- Can also double as a representation for a chunk of a value.
+data RawValue =
+    UnsafeRawValue !Int !Int ByteArray
+  | EmptyRawValue
+
+{-# COMPLETE EmptyRawValue, RawValue #-}
+pattern RawValue :: Int -> Int -> ByteArray -> RawValue
+pattern RawValue start end ba <- UnsafeRawValue start end ba
+  where RawValue start end ba =
+          assert (0 <= start && start <= end && end <= sizeofByteArray ba) $
+          UnsafeRawValue start end ba
+
+-- | Size of value in number of bytes.
+sizeofValue :: RawValue -> Int
+sizeofValue EmptyRawValue          = 0
+sizeofValue (RawValue start end _) = end - start
+
+-- | Size of value in number of bytes.
+sizeofValue16 :: RawValue -> Word16
+sizeofValue16 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue32 :: RawValue -> Word32
+sizeofValue32 = fromIntegral . sizeofValue
+
+-- | Size of value in number of bytes.
+sizeofValue64 :: RawValue -> Word64
+sizeofValue64 = fromIntegral . sizeofValue
+
+emptyValue :: RawValue
+emptyValue = EmptyRawValue
+
+--
+-- Operation
+--
+
+data Operation =
+    Insert RawValue
+  | Mupsert RawValue
+  | Delete
+
+toValue :: Operation -> RawValue
+toValue (Insert v)  = v
+toValue (Mupsert v) = v
+toValue Delete      = emptyValue
+
+opToCrumb :: Operation -> Crumb
+opToCrumb Insert{}  = 0
+opToCrumb Mupsert{} = 1
+opToCrumb Delete{}  = 2
+
+--
+-- Blob reference
+--
+
+-- | TODO: replace this placeholder type by the actual blob reference type.
+data BlobRef = BlobRef {
+    blobRefOffset :: !Word64
+  , blobRefSize   :: !Word32
+  }
+
+{-------------------------------------------------------------------------------
+  Page builder
+-------------------------------------------------------------------------------}
+
+-- | Builder with output size in bytes
+data SizedBuilder = SizedBuilder {
+    numBytes :: Word64
+  , builder  :: BB.Builder
+  }
+
+instance Semigroup SizedBuilder where
+  SizedBuilder n1 b1 <> SizedBuilder n2 b2 = SizedBuilder (n1 + n2) (b1 <> b2)
```

Comment:
I would be very surprised if this yields good performance. It almost certainly means the builder cannot be run efficiently, and has to all be allocated as a chain of closures before being executed. AFAIK, the only way to run builders efficiently is for the builder to be the _only_ result of the function. Remember it's basically a continuation.

Now in this case perhaps it doesn't matter if this is just a reference implementation and not intended to be our performant implementation.